Handle 'index out of range' errors more gracefully

When attempting to extract an entry from an empty array, the program exits rather abruptly with a panic: 

```
$ echo 'a: []' | yaml r - 'a[0]'
panic: runtime error: index out of range

goroutine 1 [running]:
main.readArray(0x3c64e0, 0x0, 0x0, 0x0, 0xc20801e670, 0x0, 0x0, 0x0, 0x0)
        /Users/morgan/Projects/golang/src/github.com/mikefarah/yaml/data_navigator.go:65 +0xd6
main.recurse(0x1b5300, 0xc20801e600, 0x7fff5fbff9e3, 0x1, 0xc20801e670, 0x0, 0x0, 0x0, 0x0)
        /Users/morgan/Projects/golang/src/github.com/mikefarah/yaml/data_navigator.go:52 +0x36e
main.calculateValue(0x1b5300, 0xc20801e600, 0xc20801e670, 0x1, 0x1, 0x0, 0x0)
        /Users/morgan/Projects/golang/src/github.com/mikefarah/yaml/data_navigator.go:80 +0xcb
main.readMap(0xc20803ca50, 0x7fff5fbff9e1, 0x1, 0xc20801e670, 0x1, 0x1, 0x0, 0x0)
        /Users/morgan/Projects/golang/src/github.com/mikefarah/yaml/data_navigator.go:25 +0x182
main.read(0xc20801e520, 0x2, 0x2, 0x0, 0x0)
        /Users/morgan/Projects/golang/src/github.com/mikefarah/yaml/yaml.go:94 +0x196
main.readProperty(0xc208088000, 0xc20801e520, 0x2, 0x2)
        /Users/morgan/Projects/golang/src/github.com/mikefarah/yaml/yaml.go:80 +0x3c
github.com/mikefarah/yaml/Godeps/_workspace/src/github.com/spf13/cobra.(*Command).execute(0xc208088000, 0xc20801e3a0, 0x2, 0x2, 0x0, 0x0)
        /Users/morgan/Projects/golang/src/github.com/mikefarah/yaml/Godeps/_workspace/src/github.com/spf13/cobra/command.go:547 +0x82f
github.com/mikefarah/yaml/Godeps/_workspace/src/github.com/spf13/cobra.(*Command).Execute(0xc2080883c0, 0x0, 0x0)
        /Users/morgan/Projects/golang/src/github.com/mikefarah/yaml/Godeps/_workspace/src/github.com/spf13/cobra/command.go:630 +0x414
main.main()
        /Users/morgan/Projects/golang/src/github.com/mikefarah/yaml/yaml.go:29 +0x3c9
```

This is a bit user-hostile, and doesn't really point to where the issue arises in the source file in order to debug it.

--- a/data_navigator.go
+++ b/data_navigator.go
@@ -58,7 +58,7 @@ func recurse(value interface{}, head string, tail []string) interface{} {
 }
 
 func readArray(array []interface{}, head int64, tail []string) interface{} {
-	if head > int64(len(array)) {
+	if head >= int64(len(array)) {
 		return nil
 	}
 

--- a/data_navigator_test.go
+++ b/data_navigator_test.go
@@ -83,6 +83,17 @@ b:
 	assertResult(t, nil, readMap(data, "b", []string{"d", "3"}))
 }
 
+func TestReadMap_with_array_out_of_bounds_by_1(t *testing.T) {
+	var data = parseData(`
+---
+b:
+  d:
+    - 3
+    - 4
+`)
+	assertResult(t, nil, readMap(data, "b", []string{"d", "2"}))
+}
+
 func TestReadMap_with_array_splat(t *testing.T) {
 	var data = parseData(`
 e:

